src/Commands/Self/Plugin/PluginBaseCommand.php

```diff
@@ -0,0 +1,276 @@
+     * @return array Plugin projects
+     * @throws TerminusNotFoundException
+     */
+    protected function getPluginProject($project)
+    {
+        $matches = array_filter(
+            $this->getPluginProjects(),
+            function($plugin_data) use ($project) {
+                return in_array($project, [$plugin_data['name'], $plugin_data['project'],]);
+            }
+        );
+        if (empty($matches)) {
+            throw new TerminusNotFoundException(self::PROJECT_NOT_FOUND_MESSAGE, compact('project'));
+        }
+        return array_shift($matches);
+    }
+
+    /**
+     * Get the plugin directory.
+     *
+     * @param string $plugin Plugin name
+     * @return string Plugin directory
+     */
+    protected function getPluginDir($plugin = '')
+    {
+        $plugins_dir = $this->getConfig()->get('plugins_dir');
+        // Create the directory if it doesn't already exist.
+        if (!is_dir("$plugins_dir")) {
+            mkdir("$plugins_dir", 0755, true);
+        }
+        return $plugins_dir . DIRECTORY_SEPARATOR . $plugin;
+    }
+
+    /**
+     * Get the plugin installation method.
+     *
+     * @param string $plugin_dir The directory the plugin is installed to
+     * @return string Plugin installation method
+     */
+    protected static function getInstallMethod($plugin_dir)
+    {
+        $git_dir = $plugin_dir . DIRECTORY_SEPARATOR . '.git';
+        if (is_dir("$git_dir")) {
+            return 'git';
+        }
+        $composer_json = $plugin_dir . DIRECTORY_SEPARATOR . 'composer.json';
+        return file_exists($composer_json) ? 'composer' : 'unknown';
+    }
+
+    /**
+     * Get the currently installed plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Installed plugin version
+     */
+    protected function getInstalledVersion($plugin)
+    {
+        exec("cd \"$plugin\" && git describe --tags", $tags);
+        if (!empty($tags)) {
+            $version = array_pop($tags);
+        } else {
+            exec("cd \"$plugin\" && git rev-parse --abbrev-ref HEAD", $branch);
+            $version = !empty($branch) ? array_pop($branch) : 'unknown';
+        }
+        return $version;
+    }
+
+    /**
+     * Get the latest available plugin version.
+     *
+     * @param string $plugin Path to plugin
+     * @return string Latest plugin version
+     */
+    protected function getLatestVersion($plugin)
+    {
+        // Get the Terminus major version.
+        $terminus_major_version = $this->getTerminusMajorVersion();
+        $cmds = "cd \"$plugin\""
+            . " && git fetch --all"
+            . " && git tag -l --sort=version:refname"
+            . " | grep ^[v$terminus_major_version] | sort -r | head -1";
+        exec($cmds, $tag);
+        if (!empty($tag)) {
+            $version = array_pop($tag);
+            // Check for non-stable semantic version (ie. -beta1 or -rc2).
+            preg_match('/(v*.*)\-(.*)/', $version, $matches);
```

Comment:
It would be nice if there was a message if a stable version cannot be found and stability = 'stable' is provided rather than just returning nothing.  Bonus: Display all the other versions available and suggest to change the stability to install.  Maybe a TODO?

Reply:
Aye, these are worthy goals and I'll add them to the TODOs.